## 005a Variant recalibration (VQSR)
This module readjusts the quality (likelihood) of variant calling, for each record in a VCF file. The aim of the module is to correct multiple testing bias.

The approach taken by variant quality score recalibration (VQRS) is to develop a continuous covarying estimate of the relationship between SNP call annotations (QD, SB, HaplotypeScore, HRun, for example) and the the probability that a SNP is a true genetic variant versus a sequencing or data processing artifact. This model is determined adaptively based on "true sites" provided as input (variant references).

The model is run in a first step to recall SNPs and then a second step to recall INDELs.

## Execution Parameters

### GATK - parameters

```

For VariantRecalibrator
-R -> path to Reference sequence file
-input -> path to one or more VCFs of raw input variants to be recalibrated
-mode -> Recalibration mode to employ (SNP or INDEL)
-recalFile -> The output recal file used by ApplyRecalibration
-tranchesFile -> The output tranches file used by ApplyRecalibration
-rscriptFile -> The output rscript file generated by the VQSR to aid in visualization of the input data and learned model
-nt -> Number of data threads to allocate to this analysis
-resource -> A list of sites for which to apply a prior probability of being correct but which aren't used by the algorithm (training and truth sets are required to run) (Mutiple resource files can be used)
-an -> The names of the annotations which should used for calculations (multiple annotations can be used)

For ApplyRecalibration
-R -> path to Reference sequence file
-input -> path to one or more VCFs of raw input variants to be recalibrated
-mode -> Recalibration mode to employ (SNP or INDEL)
-recalFile -> The output recal file used by ApplyRecalibration
-nt -> Number of data threads to allocate to this analysis
-tranchesFile -> The output tranches file used by ApplyRecalibration
-ts_filter_level -> Filter threshold. The levels of truth sensitivity at which to slice the data. (in percent, that is 1.0 for 1 percent). (AKA --TStranche)

```